File: 5-Problem-A-Day/130. Surrounded Regions.java
class Solution {
    private void check(char[][] board, int i, int j, int row, int col) {
        if (board[i][j] == 'O') {
            board[i][j] = '1';
            if (i > 1)
                check(board, i - 1, j, row, col);
            if (j > 1)
                check(board, i, j - 1, row, col);
            if (i + 1 < row)
                check(board, i + 1, j, row, col);
            if (j + 1 < col)
                check(board, i, j + 1, row, col);
        }
    }

    public void solve(char[][] board) {
        int row = board.length, col = row == 0 ? 0 : board[0].length;
        if (row == 0 || col == 0)
            return;

        for (int i = 0; i < row; i++) {
            check(board, i, 0, row, col);
            if (col > 1)
                check(board, i, col - 1, row, col);
        }

        for (int j = 1; j < col - 1; j++) {
            check(board, 0, j, row, col);
            if (row > 1)
                check(board, row - 1, j, row, col);
        }

        for (int i = 0; i < row; i++)
            for (int j = 0; j < col; j++)
                board[i][j] = board[i][j] == 'O' ? 'X' : board[i][j];

        for (int i = 0; i < row; i++)
            for (int j = 0; j < col; j++)
                board[i][j] = board[i][j] == '1' ? 'O' : board[i][j];

    }
}

File: 2023/130. Surrounded Regions.cpp
class Solution {
public:
    int dirs[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    void solve(vector<vector<char>>& board) {
        char visited = 'v';
        for (int i = 0; i < board.size(); i++) {
            if (board[i][0] == 'O') {
                bfs(board, i, 0, visited);
            }
            if (board[i][board[0].size() - 1] == 'O') {
                bfs(board, i, board[0].size() - 1, visited);
            }
        }
        for (int j = 0; j < board[0].size(); j++) {
            if (board[0][j] == 'O') {
                bfs(board, 0, j, visited);
            }
            if (board[board.size() - 1][j] == 'O') {
                bfs(board, board.size() - 1, j, visited);
            }
        }
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                if (board[i][j] == visited) {
                    board[i][j] = 'O';
                } else if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
            }
        }

    }

    void bfs(vector<vector<char>>& board, int row, int col, char visited) {
        queue<array<int, 2>> queue;
        board[row][col] = visited;
        queue.push({row, col});
        while (!queue.empty()) {
            int r = queue.front()[0], c = queue.front()[1];
            queue.pop();
            for (int i = 0; i < 4; i++) {
                int nr = r + dirs[i][0], nc = c + dirs[i][1];
                if (nr < 0 || nr >= board.size() || nc < 0 || nc >= board[0].size() || board[nr][nc] != 'O') {
                    continue;
                }
                board[nr][nc] = visited;
                queue.push({nr, nc});
            }
        }
    }
};